### PR TITLE
adds support for scan based merchant returns service

### PR DESCRIPTION
**Pull request by orenf**
https://ribbs.usps.gov/merchantreturns/documents/tech_guides/MerchantReturnsDomesticAPI.pdf

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -98,7 +99,7 @@ Active Shipping is currently being used and improved in a production environment
...
+After installing dependencies with `bundle install`, you can run the unit tests with `rake`. The unit tests mock out requests and responses so that everything runs locally, while the remote tests actually hit the carrier servers. For the remote tests, you'll need valid test credentials for any carriers' tests you want to run. The credentials should go in ~/.active_shipping/credentials.yml, and the format of that file can be seen in the included [credentials.yml](https://github.com/Shopify/active_shipping/blob/master/test/credentials.yml). For some carriers, we have public credentials you can use for testing: see `.travis.yml`.
```
Comment (kmcphillips) [inline README.md]: FYI, this will conflict with https://github.com/Shopify/active_shipping/pull/290/files#diff-04c6e90faac2675aa89e2176d2eec7d8R101

File: lib/active_shipping/carriers/usps_returns.rb [hunk 2]
```diff
@@ -0,0 +1,99 @@
+require 'cgi'
```
Comment (kmcphillips) [inline lib/active_shipping/carriers/usps_returns.rb]: If you're using CGI, can you require this at a higher level? In `lib/active_shipping.rb` or something. Be explicit about what is using it and what is being required.

File: lib/active_shipping/carriers/usps_returns.rb [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            else
+              {:error => {:code => '', :message => message}}
+            end
+          end
+        else
+          {}
+        end
+      else
+        {}
+      end
```
Comment (kmcphillips) [inline lib/active_shipping/carriers/usps_returns.rb]: This kind of else/end slide is really bad style and is difficult to read. Can you put a `result = {}` at the beginning, modify its contents, and then return it. Then you can kill all the else cases. Either that or refactor the actions inside into a small object or some other methods.

File: lib/active_shipping/carriers/usps_returns.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    def commit(action, request, test = false)
+      ssl_get(request_url(action, request, test))
+    end
+
+    def request_url(action, request, test)
+      scheme = USE_SSL[action] ? 'https://' : 'http://'
+      host = test ? TEST_DOMAINS[USE_SSL[action]] : LIVE_DOMAIN
+      resource = test ? TEST_RESOURCE : LIVE_RESOURCE
+      "#{scheme}#{host}/#{resource}?#{API_CODES[action]}=#{request}"
```
Comment (kmcphillips) [inline lib/active_shipping/carriers/usps_returns.rb]: Does `request` here need to be escaped?

File: lib/active_shipping/external_return_label_request.rb [hunk 5]
```diff
@@ -0,0 +1,479 @@
+# -*- coding: utf-8 -*-
+module ActiveShipping
+
+  class ExternalReturnLabelRequest
+
+    CAP_STRING_LEN = 100
+
+    USPS_EMAIL_REGEX = /^([a-zA-Z0-9_\-\.]+)@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.)|(([a-zA-Z0-9\-]+\.)+))([a-zA-Z]{2,4}|[0-9]{1,3})(\]?)$/
```
Comment (kmcphillips) [inline lib/active_shipping/external_return_label_request.rb]: Where did this come from?

File: lib/active_shipping/external_return_label_request.rb [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+                :sender_email,
+                :recipient_name,
+                :recipient_email,
+                :recipient_bcc,
+                :merchant_account_id,
+                :mid
+
+    def initialize(options = {})
+      options.each do |pair|
+        # TODO: Do we need to handle non-existent options
```
Comment (kmcphillips) [inline lib/active_shipping/external_return_label_request.rb]: Can we avoid `TODO`s please? They are pretty well hidden and quickly become outdated.

Fix the problem, make an issue, or raise an error.

File: lib/active_shipping/external_return_label_request.rb [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+        self.send("#{pair[0]}=".to_sym, pair[1]) if self.respond_to?("#{pair[0]}=".to_sym)
```
Comment (kmcphillips) [inline lib/active_shipping/external_return_label_request.rb]: `public_send`?

Comment (kmcphillips) [inline lib/active_shipping/external_return_label_request.rb]: And it doesn't appear so, but `options` keys are never user accessible are they? If they are, you need to whitelist the attributes or else it is dangerous.

File: test/fixtures.yml [hunk 8]
```diff
@@ -0,0 +1,44 @@
+usps:
```
Comment (kmcphillips) [inline test/fixtures.yml]: I don't understand how this file is used? Where did all this information come from?

**Comment by kmcphillips** [comment]
A few questions and comments. Mostly good work. If you can address them and get the build green and we'll get this merged soon for you.

Thank you for your contribution.

File: lib/active_shipping/carriers/usps_returns.rb [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+    self.retry_safe = true
+
+    cattr_reader :name
+    @@name = "USPS Returns"
+
+    LIVE_DOMAIN = 'returns.usps.com'
+    LIVE_RESOURCE = 'Services/ExternalCreateReturnLabel.svc/ExternalCreateReturnLabel'
+
+    TEST_DOMAINS = { #indexed by security; e.g. TEST_DOMAINS[USE_SSL[:rates]]
```
Comment (trishume) [inline lib/active_shipping/carriers/usps_returns.rb]: Why is this here if both possibilities are the same?

File: lib/active_shipping/carriers/usps_returns.rb [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    USE_SSL = {
+      :external_return_label_request => true
+    }
+
+    def requirements
+      []
+    end
+
+    def external_return_label_request(label, options = {})
+      response = commit(:external_return_label_request, URI.encode(label.to_xml), (options[:test] || false))
```
Comment (trishume) [inline lib/active_shipping/carriers/usps_returns.rb]: As @kmcphillips was kind of getting at but missed this part. The URI encoding should really be done in the `request_url` method, not here.

File: README.md (hunk 1 above)
Comment (orenf) [inline README.md]: Yeah, sorry, I modified this text because it doesn't look like you have a task called ```test:units``` or  ```test:remote```. I will revert, however.

File: lib/active_shipping/carriers/usps_returns.rb (hunk 2 above)
Comment (orenf) [inline lib/active_shipping/carriers/usps_returns.rb]: yes, will change.

File: lib/active_shipping/carriers/usps_returns.rb (hunk 3 above)
Comment (orenf) [inline lib/active_shipping/carriers/usps_returns.rb]: Yes, will fix.

File: README.md (hunk 1 above)
Comment (trishume) [inline README.md]: That's exactly what the other PR noticed and fixed. :stuck_out_tongue: 
The other one is smaller and closer to being merged though.

File: lib/active_shipping/external_return_label_request.rb (hunk 5 above)
Comment (orenf) [inline lib/active_shipping/external_return_label_request.rb]: Straight from the usps API docs.

File: lib/active_shipping/external_return_label_request.rb (hunk 6 above)
Comment (orenf) [inline lib/active_shipping/external_return_label_request.rb]: def will fix.

File: test/fixtures.yml (hunk 8 above)
Comment (orenf) [inline test/fixtures.yml]: My apologies. This code has been running in production for the better part of a year. I only recently rebased, and this got added back in. Will fix.

File: lib/active_shipping/carriers/usps_returns.rb (hunk 9 above)
Comment (orenf) [inline lib/active_shipping/carriers/usps_returns.rb]: Will fix.

File: lib/active_shipping/carriers/usps_returns.rb (hunk 10 above)
Comment (orenf) [inline lib/active_shipping/carriers/usps_returns.rb]: Will fix. Thank you.

**Comment by orenf** [comment]
@kmcphillips and @trishume thank you so much for your feedback. I really appreciate it.

**Comment by orenf** [comment]
Still waiting on test credentials that I can pass on from USPS in order to get the build green.  Otherwise, I believe this pull includes all the changes you suggested.  @kmcphillips what is the typical process here for running remote tests that require credentials?

**Comment by trishume** [comment]
This is currently a pain point in active_shipping. Our CI only includes a small set of credentials which we have obtained and know are safe to release publicly. We're working on expanding that set but for now most of the tests are set to skip on authentication failure and I just run the remote tests locally before I do a release with Shopify's test credentials set.

I'm fine with merging this without public credentials provided you could point me in the right direction for me to get some for Shopify's internal use. If you gave us a set of credentials along with this PR that would be amazing and you would join the list of my favorite contributors ever :heart:

File: lib/active_shipping/carriers/usps_returns.rb [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+      )
+    end
+
+    def external_return_label_errors(document)
+      return {} unless document.respond_to?(:elements)
+      res = {}
+      if node = document.at('*/errors')
+        if node.at('ExternalReturnLabelError')
+          if message = node.at('ExternalReturnLabelError/InternalErrorDescription').try(:text)
+            if code = node.at('ExternalReturnLabelError/InternalErrorNumber').try(:text)
```
Comment (trishume) [inline lib/active_shipping/carriers/usps_returns.rb]:
```ruby
code = node.at('ExternalReturnLabelError/InternalErrorNumber').try(:text) || ''
res = {:error => {:code => code, :message => message}}
```
and then get rid of the if statement, makes this method way simpler. Same for the other branch.

File: lib/active_shipping/external_return_label_request.rb [hunk 12]
```diff
@@ -0,0 +1,478 @@
...
+      if v.is_a?(String)
+        v.strip!
+        v[0..CAP_STRING_LEN - 1]
+      else
+        nil
+      end
+    end
+
+    # Sent by the system containing the returns label attachment and message.
+    def recipient_bcc=(v)
```
Comment (trishume) [inline lib/active_shipping/external_return_label_request.rb]: I love the level of validation in this, so much friendlier than the rest of Active Shipping where you get back some obscure error message that has nothing to do with what you did wrong. Great work finding out and coding the requirements for all of these fields. :+1: 

And don't even get me started on the doc strings describing what each field actually means, miles beyond the opaque tag names everywhere else in the codebase. So many times I've had to track down postal service docs just to figure out what a tag is. :heart_eyes_cat: 

Lots of copy-paste though. I think these could all be refactored to use some private `validate_x` methods. I think `validate_email(s)`, `validate_string_length(s,max_len)`, `validate_range(n,min,max)`, `validate_included(v,options)` that test a value and then raise a good error would cover everything and then each method could just be like
```ruby
def recipient_email=(v)
  v = sanitize(v)
  validate_email(v)
  @recipient_email = v
end
```

File: test/fixtures/xml/usps/invalid_xml_response.xml [hunk 13]
```diff
@@ -0,0 +1,2 @@
+<ExternalReturnLabelErrorResponse xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"><errors><ExternalReturnLabelError><InternalErrorNumber>4001</InternalErrorNumber><InternalErrorDescription>Authorization
```
Comment (trishume) [inline test/fixtures/xml/usps/invalid_xml_response.xml]: It would be nice if these were formatted.

**Comment by trishume** [comment]
Just realized I hadn't actually reviewed this yet. Found some coding style things that could do with some refactoring.

This PR is just outstanding though, the level of documentation, validation, testing and everything else blows me away. And the fact that you without even being prompted say you're looking into getting us credentials. And the fact that you say it's been tested in production for a long time. Just wow.

In fact since this PR is so great I'm willing to do the refactors I suggested myself if you'd like. I have pro editor macro skills so it shouldn't take long.

File: lib/active_shipping/carriers/usps_returns.rb (hunk 11 above)
Comment (orenf) [inline lib/active_shipping/carriers/usps_returns.rb]: This is a no brainer. I already fixed. Thanks for your suggestion.

File: test/fixtures/xml/usps/invalid_xml_response.xml (hunk 13 above)
Comment (orenf) [inline test/fixtures/xml/usps/invalid_xml_response.xml]: You got it. fixed.

**Comment by orenf** [comment]
@trishume thank you for your very kind words !!!

I think you're right that I can and should refactor some of the validations.  I need a couple of min to take care of those changes. 
In the meantime, I just pushed up fixes for other suggestions you made. 

Also, I realized that remote tests were rescue'ing with NoCredentialsFound, which is something I failed to do. And lastly, I'm still waiting on a response from USPS. I received an e-mail earlier that they were looking into getting me some creds.

**Comment by orenf** [comment]
@trishume , I applied your recommendations for dry'ing up validations. I had the validations themselves executing ```sanitize```. Not sure how you feel about that. Anyway, let me know if there is something I missed. I've added a few more tests, and they all seem to be passing.

**Comment by trishume** [comment]
Amazing :+1: 

I'll merge this when I do the next release. If USPS gets back to you with credentials and you can make them public you can submit another PR later (if I've already merged this one) or if they can't be public but you'd like to donate them to Shopify you can email shipping@shopify.com.

**Comment by orenf** [comment]
Thank you @trishume, I sent another e-mail to USPS to request temporary credentials for the sake of testing that you can do on your end.  I'll continue to follow up with them and will let you know accordingly.

**Comment by trishume** [comment]
Sure that would be great.

**Comment by orenf** [comment]
@trishume 
1. Squashed
2. I have USPS credentials. For this feature, there is no test server, it's all production. I was able to get a test account set up with a very stern warning not to make it public. Additionally, the account will be turned off by end of the month.  Can I e-mail shipping@shopify.com and maintain correspondence with a single person to pass along the details of this account and ensure that this feature works against remote?

**Comment by trishume** [comment]
Great. shipping@shopify.com goes to the entire shipping team. If you want to have a specific contact I'm tristan.hume@shopify.com but I'm an intern and will only be here for 2 more months. If you want a more permanent contact kevin.mcphillips@shopify.com is the shipping team lead.
